##### Q1) From a list of words: ["apple","mango","kiwi","banana","pear"]
Filter only those words whose length is greater than 4.

In [12]:
ls = ["apple","mango","kiwi","banana","pear"]
[x for x in ls if len(x)>4]

['apple', 'mango', 'banana']

##### Q2) Given list: [10,15,20,25,30,35]
Filter only numbers greater than 20 and divisible by 5 (use two if conditions).

In [13]:
ls = [10,15,20,25,30,35]
[x for x in ls if x>20 if x%5==0]

[25, 30, 35]

##### Q3) Given list [1,2,3,4,5], create a new list where
even numbers become "EVEN"
odd numbers become "ODD"
Using list comprehension with X if cond else Y.

In [14]:
ls = [1,2,3,4,5]
["Even" if x%2==0 else "Odd" for x in ls]

['Odd', 'Even', 'Odd', 'Even', 'Odd']

##### Q4) Given ["red","blue","red","green"],
Replace "red" with "R" otherwise keep the same word.

ls = ["red","blue","red","greem"]
["R" if x=="red" else x for x in ls]

##### Given [5,10,15,20],
Replace numbers > 10 with "BIG" and numbers ≤ 10 with "SMALL".

ls = [5,10,15,20]
["BIG" if x >10 else "SMALL" for x in ls]

##### From numbers 0–20, select numbers that are:

divisible by 2

divisible by 3
(two if conditions)

In [15]:
[x for x in range(21) if x%2==0 if x%3==0]

[0, 6, 12, 18]

##### From list [12, 5, 8, 20, 7, 30],
Filter numbers:
greater than 10
even
less than 25
(use 3 chained if conditions)

In [16]:
ls = [12,5,8,20,7,30]
[x for x in ls if x%2==0 if 25>x>10]

[12, 20]

##### Given list of names:
["john","bob","michael","adam","alex"]
Filter names whose:

length > 3

starts with "a"
(two chained conditions)

In [17]:
ls = ["john","bob","michael","adam","alex"]
[x for x in ls if len(x) if x[0].lower()=="a"]

['adam', 'alex']

##### Given [1,2,3,4,5,6,7,8],
Create a new list where:

if the number is even → store "E"

else → store "O"
BUT ONLY for numbers greater than 3 and less than 7
(use expression + two chained conditions)

In [18]:
lsb= [1,2,3,4,5,6,7,8]
["E" if x%2==0 else "O" for x in lsb if 7>x>3 ]

['E', 'O', 'E']

##### Q11:

Given a sentence:
"Python is easy and powerful"
Create a list of words having:

length ≥ 4

not containing letter "a"
(use chained if conditions)

In [19]:
s = "Python is easy and powerful"
[x for x in s.split() if len(x)>=4 if "a" not in x]

['Python', 'powerful']

##### Q12:

Given [10, 25, 37, 40, 55, 60],
Construct a new list with:

"YES" if number divisible by 5 and 2

"NO" otherwise
(using expression → X if cond else Y)

In [20]:
ls = [10,25,37,40,55,60]
["YES" if x%2==0 and x%5==0 else "NO" for x in ls]

['YES', 'NO', 'NO', 'YES', 'NO', 'YES']

##### Q13:

Create a list of squares of numbers from 1–20 BUT include only:

odd numbers

and numbers ending in digit 3 or 5
(E.g., 3,5,13,15 → valid)

In [1]:
[x for x in range(1,21) if x%2!=0 if (x % 10)==3 or (x % 10)==5]

[3, 5, 13, 15]

Given:
matrix = [[2, 4, 6], [1, 3], [5]]

Flatten the matrix into:

[2, 4, 6, 1, 3, 5]

using double for inside a single list comprehension.

In [2]:
matrix = [[2, 4, 6], [1, 3], [5]]
[x for y in matrix for x in y]

[2, 4, 6, 1, 3, 5]

Given:

words = [["hi", "hello"], ["cat"], ["python", "rocks"]]


Flatten the list and convert every word to uppercase in one line.

In [3]:
words = [["hi", "hello"], ["cat"], ["python", "rocks"]]
[x.upper() for y in words for x in y]

['HI', 'HELLO', 'CAT', 'PYTHON', 'ROCKS']

Create a 4×5 matrix of zeros using list comprehension.

In [4]:
[[0 for x in range(5)] for y in range(4)]

[[0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0], [0, 0, 0, 0, 0]]

matrix = [[1, 2, 3, 4],[5, 6, 7, 8]]
Create the transpose using nested list comprehension.

In [5]:
matrix = [
    [1, 2, 3, 4],
    [5, 6, 7, 8]]
[[x[y] for x in matrix] for y in range(len(matrix[0]))]

[[1, 5], [2, 6], [3, 7], [4, 8]]

m = [
    [3, 1],
    [2, 4]
]


Create a new matrix which stores the square of each element using nested comprehension.

In [6]:
m = [[3,1],[2,4]]
[[y**2 for y in x]for x in m ]

[[9, 1], [4, 16]]

matrix = 
    [[10, 20],
    [30, 40],
    [50, 60]]
Extract the first column only using list comprehension.
Expected:
[10, 30, 50]

In [7]:
matrix =[[10, 20],[30, 40],[50, 60]]
[x[y] for x in matrix for y in range(1)]
[x[0] for x in matrix]

[10, 30, 50]

names = ["john", "ALICE", "mArK"]


Generate a new list where:

first letter capitalized

remaining letters lowercase
(but do it without using capitalize()—use slicing inside comprehension)

In [8]:
names = ["john", "ALICE", "mArK"]
[x[0].upper()+x[1:].lower() for x in names]

['John', 'Alice', 'Mark']

Using math.sqrt() and round(), generate a list of square roots of numbers 1 to 10, rounded to 3 decimals.

In [9]:
import math
[round(math.sqrt(x),3) for x in range(1,11)]

[1.0, 1.414, 1.732, 2.0, 2.236, 2.449, 2.646, 2.828, 3.0, 3.162]

nums = [-5, 10, -15, 20]


Return "POS" if number positive else "NEG" and store as:

["NEG", "POS", "NEG", "POS"]


using conditional expression + f-string (optional use).

In [10]:
nums = [-5,10,-15,20]
["POS" if x>0 else "NEG" for x in nums]

['NEG', 'POS', 'NEG', 'POS']

Q10.

Create a function cube(x)
Then use list comprehension to generate:

[1, 8, 27, 64, ...]  # cubes from 1 to 10

In [11]:
def sqr(x):
    return x**3
[sqr(x) for x in range(1,11)]

[1, 8, 27, 64, 125, 216, 343, 512, 729, 1000]

Given function:
def reverse_string(s):
    return s[::-1]
Use list comprehension to reverse each word in:
["apple", "banana", "cat"]

In [12]:
def reverse_string(s):
    return s[::-1]
ls = ["apple", "banana", "cat"]
[reverse_string(x) for x in ls]

['elppa', 'ananab', 'tac']

Given:

def classify(n):
    return "Even" if n % 2 == 0 else "Odd"

Use list comprehension to classify all numbers from 1 to 15.

In [13]:
def classify(n):
    return "Even" if n%2==0 else "Odd"
[classify(x) for x in range(1,16)]

['Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd',
 'Even',
 'Odd']

Flatten this 2D list AND compute square of each element:

data = [[1, 2], [3, 4, 5], [6]]


Expected:

[1, 4, 9, 16, 25, 36]

In [14]:
data = [[1,2],[3,4,5],[6]]
[x**2 for y in data for x in y ]

[1, 4, 9, 16, 25, 36]

matrix = [[5, 10],[15, 20],[25, 30]]

Create new list of strings:

["5->25", "10->100", ...]

(each element mapped to element→element²)

In [15]:
matrix = [[5, 10],[15, 20],[25, 30]]
[f"{x}->{x**2}" for y in matrix for x in y]

['5->25', '10->100', '15->225', '20->400', '25->625', '30->900']

Using nested comprehension + f-string + math:
Create:

["sqrt(1) = 1.00", "sqrt(2) = 1.41", ... sqrt(10)]

In [16]:
import math
[f"{x}={round(math.sqrt(x),2)}" for x in range(1,11)]

['1=1.0',
 '2=1.41',
 '3=1.73',
 '4=2.0',
 '5=2.24',
 '6=2.45',
 '7=2.65',
 '8=2.83',
 '9=3.0',
 '10=3.16']

Write list comprehension to transpose ANY matrix:

matrix = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]]

In [17]:
matrix = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]]
[[x[y] for x in matrix] for y in range(len(matrix[0]))]

[[1, 4, 7], [2, 5, 8], [3, 6, 9]]

matrix = [["a1","a2"],
    ["b1","b2"],
    ["c1","c2"]]

convert this matrix into into:

["a1-b1-c1", "a2-b2-c2"]

using nested loops + join inside comprehension.


In [18]:
matrix = [["a1","a2"],["b1","b2"],["c1","c2"]]
["-".join([f"{x[y]}" for x in matrix]) for y in range(len(matrix[0]))]

['a1-b1-c1', 'a2-b2-c2']

Q18.

Given:

students = [
    {"name": "Aman", "scores": [80, 90, 70]},
    {"name": "Sara", "scores": [88, 76, 92]}
]
Extract only the average score for each student using:
nested loops
sum()
len()
list comprehension

In [19]:
students = [
    {"name": "Aman", "scores": [80, 90, 70]},
    {"name": "Sara", "scores": [88, 76, 92]}
]
[[f"{y}" for y in x.values()] for x in students]

[['Aman', '[80, 90, 70]'], ['Sara', '[88, 76, 92]']]

sentences = ["Python is great", "List comprehension is powerful"]


Create a list containing all words, flattened

In [20]:
sentences = ["Python is great", "List comprehension is powerful"]
[y for x in sentences for y in x if y!=" "]

['P',
 'y',
 't',
 'h',
 'o',
 'n',
 'i',
 's',
 'g',
 'r',
 'e',
 'a',
 't',
 'L',
 'i',
 's',
 't',
 'c',
 'o',
 'm',
 'p',
 'r',
 'e',
 'h',
 'e',
 'n',
 's',
 'i',
 'o',
 'n',
 'i',
 's',
 'p',
 'o',
 'w',
 'e',
 'r',
 'f',
 'u',
 'l']

matrix = [
    [1, -2, 3],
    [-4, 5, -6]
]


Convert it into a matrix of absolute values using nested comprehension.

In [21]:
matrix = [
    [1, -2, 3],
    [-4, 5, -6]
]
[[0 for y in x ]for x in matrix ]

[[0, 0, 0], [0, 0, 0]]